dictation/dictation_context.py

```diff
@@ -0,0 +1,141 @@
+import traceback
+from dataclasses import dataclass
+from typing import Optional
+
+from talon import Context, ui, Module, actions
+from talon.mac.ui import Element
+from talon.types import Span
+
+ctx = Context()
+ctx.matches = "os: mac"
+
+mod = Module()
+setting_accessibility_dictation = mod.setting(
+    "accessibility_dictation",
+    type=bool,
+    default=False,
+    desc="Use accessibility APIs to implement context aware dictation.",
+)
+
+# Default number of characters to use to acquire context. Somewhat arbitrary.
+# The current dictation formatter doesn't need very many, but that could change in the future.
+DEFAULT_CONTEXT_CHARACTERS = 30
+
+
+@dataclass
+class AccessibilityContext:
+    """Records the context needed for dictation"""
+    content: str
+    selection: Span
+
+    def left_context(self, num_chars: int = DEFAULT_CONTEXT_CHARACTERS) -> str:
+        """Returns `num_chars`' worth of context to the left of the cursor"""
+        start = max(0, self.selection.left - num_chars)
+        return self.content[start:self.selection.left]
+
+    def right_context(self, num_chars: int = DEFAULT_CONTEXT_CHARACTERS) -> str:
+        """Returns `num_chars`' worth of context to the right of the cursor"""
+        end = min(self.selection.right + num_chars, len(self.content))
+        return self.content[self.selection.right:end]
+
+
+@mod.action_class
+class ModActions:
+
+    def accessibility_dictation_enabled():
+        """Just exports `setting_accessibility_dictation` for use in other files"""
+        return setting_accessibility_dictation.get()
+    
+    def dictation_current_element() -> Element:
+        """Returns the accessibility element that should be use for dictation (i.e.
+        the current input textbox).
+        
+        This is always always the focused (current) element, however, this function
+        exists so that context can overwrite it, for applications with strange behavior.
+        """
+        return ui.focused_element()
+
+    def accessibility_adjust_context_for_application(el: Element,
+                                                     context: AccessibilityContext) -> AccessibilityContext:
+        """Sometimes the accessibility context reported by the application is wrong, but fixable
+        in predictable ways (this is most common in Electron apps).
+        
+        This method can be overwritten in those applications to do so.
+        
+        TODO(pcohen): it's a it strange to have both this and dictation_current_element;
+        possibly refactor.
+        """
+        return context
+
+    def accessibility_create_dictation_context(el: Element) -> Optional[AccessibilityContext]:
+        """Creates a `AccessibilityContext` representing the state of the input buffer
```

Comment:
Have you checked that these multiline action docstrings are OK? Happy to have more verbose documentation but they're longer than in most of knausj.

Reply:
hm, in what way could they cause problems? Guess they do violate PEP

Reply:
Just wondering if they would cause problems with things like actions.list/actions.find, etc. Might be worth asking aegis if it’s unclear.

Reply:
ah good point. I'll either shorten it or put it on one line.